proposals/p1885.md

```diff
@@ -0,0 +1,417 @@
+(though specific):
+
+```rust
+struct Fibonacci {
+    curr: u32,
+    next: u32,
+}
+
+impl Iterator for Fibonacci {
+    type Item = u32;
+    fn next(&mut self) -> Option<Self::Item> {
+        // calculate and yield value
+        [...]
+    }
+}
+
+for i in fibonacci().take(4) {
+    ...
+}
+```
+
+#### Typescript
+
+Typescript uses the
+[`.iterator` property](https://www.typescriptlang.org/docs/handbook/iterators-and-generators.html#iterables),
+along with
+[two forms](https://www.typescriptlang.org/docs/handbook/iterators-and-generators.html#forof-statements)
+of range-based for loops: `for..in` and `for..of`.
+
+```ts
+let pets = new Set(['Cat', 'Dog', 'Hamster']);
+pets['species'] = 'mammals';
+for (let pet in pets) {
+    console.log(pet); // "species"
+}
+for (let pet of pets) {
+    console.log(pet); // "Cat", "Dog", "Hamster"
+}
+```
+
+Using the `.iterator` property, Typescript allow to lazily generate data easily
+using functions, classes, or objects:
+
+```ts
+const reverse = (arr) => ({
+    [Symbol.iterator]() {
+        let i = arr.length;
+        return {
+            next: () => ({
+                value: arr[--i],
+                done: i < 0,
+            }),
+        };
+    },
+});
+```
+
+#### Go
+
+Go has non-standard support for range-based with custom types only.
+
+https://stackoverflow.com/questions/35810674/in-go-is-it-possible-to-iterate-over-a-custom-type
+
+## Proposal
+
+Provide interfaces that can be implemented by user types to enable support for
+ranged-for loops.
+
+We propose a new [**Iterable**](#iterable) interface, for iterator-based
+sequential access, and usage with range-based `for`.
+
+## Details
+
+This proposal exposes an Iterable-type interface, relying on 2 new interfaces:
+`Iterable` and `Iterator`.
+
+A user implements these interfaces to enable range-based `for` support. Carbon
+would make use of the `Iterable` to retrieve an `Iterator` and iterate until
+completed or interrupted.
+
+### Iterable
+
+The `Iterable` interface represents a generic sequential-access type of
+container. It relies on two new interfaces:
+
+-   `Iterable`: Interface for a type that can return an iterator and be iterated
+    over
+-   `Iterator`: Interface for an iterator-like object
+
+**Note:** This interface makes use of `Optional` type, and assumes its future
+definition. The usage highlighted here is hypothetical.
+
+#### Range-based `for` behavior
+
+-   Get an iterator with `GetIterator()`
+-   For each iteration, call `Next() -> Optional(ElementType)` on the iterator
+    -   If the `Optional` is null, break the loop. Otherwise return the value
+        held by the Optional.
+
+#### Interface and usage
+
+```carbon
+// Basic immutable iterator
+interface Iterator {
+    let ElementType:! Type;
+    // Needed in the broader context of Iterators and C++ interop
+    fn Get[me: Self]() -> Optional(ElementType);
+    fn Next[addr me: *Self]() -> Optional(ElementType);
```

Comment:
Could we define an `Advance` method that moves the iterator forward but doesn't return a value, and then give `Next` a `final` definition in terms of `Advance` and `Get`? That seems more orthogonal and easier to implement, and exposing the ability to increment without the overhead of returning a value could be a useful optimization in some cases.

Reply:
Agree, done